spec/index.bs

```diff
@@ -1835,27 +1832,83 @@ dictionary GPURenderBundleEncoderDescriptor : GPUObjectDescriptorBase {
+dictionary GPUTextureDataLayout {
+    GPUBufferSize offset = 0;
+    required unsigned long rowPitch;
+    required unsigned long imageHeight;
+};
+</script>
+
 <script type=idl>
 interface GPUQueue {
     void submit(sequence<GPUCommandBuffer> commandBuffers);
 
     GPUFence createFence(optional GPUFenceDescriptor descriptor = {});
     void signal(GPUFence fence, unsigned long long signalValue);
 
+    void writeBuffer(
+        ArrayBuffer sourceData,
+        GPUBufferSize sourceOffset,
+        GPUBuffer destination,
+        GPUBufferSize destinationOffset,
+        GPUBufferSize size);
+
+    void writeTexture(
+        ArrayBuffer sourceData,
+        GPUTextureDataLayout sourceLayout,
+        GPUTextureCopyView destination,
+        GPUExtent3D size);
+
     void copyImageBitmapToTexture(
         GPUImageBitmapCopyView source,
         GPUTextureCopyView destination,
-        GPUExtent3D copySize);
+        GPUExtent3D size);
 };
 GPUQueue includes GPUObjectBase;
 </script>
 
- - {{GPUQueue/copyImageBitmapToTexture()}}:
-   - For now, `copySize.z` must be `1`.
+<dl dfn-type="method" dfn-for="GPUQueue">
+    : <dfn>writeBuffer(sourceData, sourceOffset, destination, destinationOffset, size)</dfn>
+    ::
+        Takes the `sourceData` contents starting from the offset `sourceOffset` and
+        schedules a write operation of these contents to the `destination` buffer in the queue.
+        The operation does nothing and produces an error if any of the following is true:
+
+         - `destination` buffer wasn't created with {{GPUBufferUsage/COPY_DST}} flag in {{GPUBufferDescriptor/usage}}.
+         - `destination` buffer must not be destroyed
+         - `destinationOffset + size` exceeds the {{GPUBufferDescriptor/size}} of the `destination` buffer.
+         - `destination` buffer isn't in the `"unmapped"` [=buffer state=].
+         - `destinationOffset` is not a multiple of 4.
+         - `size` is not a positive multiple of 4.
+         - `sourceOffset + size` exceeds `sourceData.byteLength`.
+
+    : <dfn>writeTexture(sourceData, sourceLayout, destination, size)</dfn>
+    ::
+        Takes the `source` contents and schedules a write operation of these contents to the
+        `destination` texture copy view in the queue.
+        The operation does nothing and produces an error if any of the following is true:
+
+         - `destination.texture` texture wasn't created with {{GPUTextureUsage/COPY_DST}} flag in {{GPUTextureDescriptor/usage}}.
+         - `destination.texture` must not be destroyed
+         - `destination.origin + size` exceeds the size of texture dimensions of the mipmap `destination.mipLevel`.
+         - any of the `size` extents are zero
+         - `sourceOffset + requiredSize` exceeds `sourceData.byteLength`, where `requiredSize` can be computed as
+            `(sourceLayout.imageHeight * sourceLayout.rowPitch * (size.depth - 1) + sourceLayout.rowPitch * (size.height - 1) + size.width) * bpp`.
```

Comment:
I presume that `rowPitch` is specified as being bytes per scanline, regardless of the pixel format.  If so, we shouldn't also be multiplying everything by `bpp`